src/_P009_MCP.ino

```diff
@@ -67,31 +124,151 @@ boolean Plugin_009(byte function, struct EventStruct *event, String& string)
+        if there is a 'state' change, check debounce period.
+        Then if doubleclick interval exceeded, reset Settings.TaskDevicePluginConfig[event->TaskIndex][7] to 0
+        Settings.TaskDevicePluginConfig[event->TaskIndex][7] contains the current status for doubleclick:
+        0: start counting
+        1: 1st click
+        2: 2nd click
+        3: 3rd click = doubleclick event if inside interval (calculated as: '3rd click time' minus '1st click time')
+
+        Returned EVENT value is = 3 always for doubleclick
+        In rules this can be checked:
+        on Button#Switch=3 do //will fire if doubleclick
+        \**************************************************************************/
+        if (state != -1 && state != switchstate[event->TaskIndex])
         {
-          if (state != switchstate[event->TaskIndex])
+          Settings.TaskDevicePluginConfigLong[event->TaskIndex][2]=millis();
+          Settings.TaskDevicePluginConfig[event->TaskIndex][6] = false;
+
+          const unsigned long debounceTime = timePassedSince(Settings.TaskDevicePluginConfigLong[event->TaskIndex][0]);
+          if (debounceTime >= (unsigned long)lround(Settings.TaskDevicePluginConfigFloat[event->TaskIndex][0])) //de-bounce check
           {
+            const unsigned long deltaDC = timePassedSince(Settings.TaskDevicePluginConfigLong[event->TaskIndex][1]);
+            if (deltaDC >= (unsigned long)lround(Settings.TaskDevicePluginConfigFloat[event->TaskIndex][1]))
+            {
+              //reset timer for doubleclick
+              Settings.TaskDevicePluginConfig[event->TaskIndex][7]=0;
+              Settings.TaskDevicePluginConfigLong[event->TaskIndex][1]=millis();
+            }
+            Settings.TaskDevicePluginConfig[event->TaskIndex][7]++;
+
+            switchstate[event->TaskIndex] = state;
+
+            byte output_value;
+            boolean sendState = switchstate[event->TaskIndex];
+
+            if (Settings.TaskDevicePin1Inversed[event->TaskIndex])
+              sendState = !sendState;
+
+            if (Settings.TaskDevicePluginConfig[event->TaskIndex][7]==3 && Settings.TaskDevicePluginConfig[event->TaskIndex][4])
+            {
+              output_value = 3; //double click
+            } else {
+              output_value = state ? 1 : 0; //single click
+            }
+
+            UserVar[event->BaseVarIndex] = output_value;
+
             String log = F("MCP  : State ");
             log += state;
+            log += output_value<=1 ? F(" Output value=") : F(" Doubleclick=");
+            log += output_value;
             addLog(LOG_LEVEL_INFO, log);
-            switchstate[event->TaskIndex] = state;
-            UserVar[event->BaseVarIndex] = state;
+
             event->sensorType = SENSOR_TYPE_SWITCH;
             sendData(event);
+
+            //reset Userdata so it displays the correct state value in the web page
+            UserVar[event->BaseVarIndex] = sendState ? 1 : 0;
```

Comment:
I get it why you want to reset the UserVar value.
I don't know if the event in `sendData` is being prepared fully before exit. (collecting all values for further processing to controllers)
Maybe worth to check if it is indeed being processed (stored in the controller delay queue) or else it will be very hard to debug errors later.

Reply:
Good point. But I need your help in how to check if sendData